src/spi/blocking.rs

```diff
@@ -1,108 +1,287 @@
+//! SPI MCU peripheral, plus the SCK, MOSI and MISO pins.
+//!
+//! Owning an instance of an SPI bus guarantees exclusive access, this is, we have the guarantee no other
+//! piece of code will try to use the bus while we own it.
+//!
+//! There's 3 bus traits, depending on the bus capabilities.
+//!
+//! - [`SpiBus`]: Read-write access. This is the most commonly used.
+//! - [`SpiBusRead`]: Read-only access, for example a bus with a MISO pin but no MOSI pin.
+//! - [`SpiBusWrite`]: Read-write access, for example a bus with a MOSI pin but no MISO pin.
+//!
+//! ## Device
+//!
+//! [`SpiDevice`] represents **ownership over a single SPI device selected by a CS pin** in a (possibly shared) bus. This is typically:
+//!
+//! - Exclusive ownership of the **CS pin**.
+//! - Access to the **underlying SPI bus**. If shared, it'll be behind some kind of lock/mutex.
+//!
+//! An [`SpiDevice`] allows initiating [transactions](SpiDevice::transaction) against the target device on the bus. A transaction
+//! consists of asserting CS, then doing one or more transfers, then deasserting CS. For the entire duration of the transaction, the [`SpiDevice`]
+//! implementation will ensure no other transaction can be opened on the same bus. This is the key that allows correct sharing of the bus.
+//!
+//! The capabilities of the bus (read-write, read-only or read-write) are determined by which of the [`SpiBus`], [`SpiBusRead`] [`SpiBusWrite`] traits
+//! are implemented for the [`Bus`](SpiDevice::Bus) associated type.
+//!
+//! # For driver authors
+//!
+//! When implementing a driver, it's crucial to pick the right trait, to ensure correct operation
+//! with maximum interoperability. Here are some guidelines depending on the device you're implementing a driver for:
+//!
+//! If your device **has a CS pin**, use [`SpiDevice`]. Do not manually manage the CS pin, the [`SpiDevice`] implementation will do it for you.
+//! Add bounds like `where T::Bus: SpiBus`, `where T::Bus: SpiBusRead`, `where T::Bus: SpiBusWrite` to specify the kind of access you need.
+//! By using [`SpiDevice`], your driver will cooperate nicely with other drivers for other devices in the same shared SPI bus.
+//!
+//! ```
+//! # use embedded_hal::spi::blocking::{SpiBus, SpiBusRead, SpiBusWrite, SpiDevice};
+//! pub struct MyDriver<SPI> {
+//!     spi: SPI,
+//! }
+//!
+//! impl<SPI> MyDriver<SPI>
+//! where
+//!     SPI: SpiDevice,
+//!     SPI::Bus: SpiBus, // or SpiBusRead/SpiBusWrite if you only need to read or only write.
+//! {
+//!     pub fn new(spi: SPI) -> Self {
+//!         Self { spi }
+//!     }
+//!
+//!     pub fn read_foo(&mut self) -> [u8; 2] {
+//!         let mut buf = [0; 2];
+//!
+//!         // `transaction` asserts and deasserts CS for us. No need to do it manually!
+//!         self.spi.transaction(|bus| {
+//!             bus.write(&[0x90]).unwrap();
+//!             bus.read(&mut buf).unwrap();
```

Comment:
```suggestion
//!             bus.write(&[0x90])?;
//!             bus.read(&mut buf)
```
Can we rather do this in this and the rest of examples?

Reply:
Done